_posts/2024-04-10-benchmarks-over-time.md

```diff
@@ -0,0 +1,548 @@
+        .then(res => res.json())
+        .then(parsed_json => {
+            let perf_over_time_scale_by_time = document.getElementById('perf_over_time_scale_by_time');
+            Plotly.plot( perf_over_time_scale_by_time, parsed_json.data, parsed_json.layout );
+            });
+</script>
+
+In version 0.9.0, launched in September 2023, [DuckDB's hash aggregate was enhanced to handle out-of-core (larger than memory) intermediates](https://github.com/duckdb/duckdb/pull/7931).
+The details of the algorithm, along with some benchmarks, are available in [this blog post](https://duckdb.org/2024/03/29/external-aggregation.html). 
+This allows for DuckDB to aggregate one billion rows of data (50GB in size) on a MacBook Pro with only 16GB of RAM, even when the number of unique groups in the group by is large. 
+This represents at least a 10x improvement in aggregate processing scale over the course of the 3 years of the benchmark.
+
+DuckDB's hash join operator has supported larger-than-memory joins since version 0.6.1. 
+However, the scale of this benchmark (coupled with the limited RAM of the benchmarking hardware), meant that this benchmark could still not complete successfully.
+In version 0.10.0, launched in February 2024, [DuckDB's memory management received a significant upgrade](https://github.com/duckdb/duckdb/pull/10147) to handle multiple concurrent operators all requiring significant memory. 
+The [0.10.0 release blog post](https://duckdb.org/2024/02/13/announcing-duckdb-0100.html#temporary-memory-manager) shares additional details about this feature. 
+
+As a result, by version 0.10.0 DuckDB was able to handle calculations on data that is significantly larger than memory, even if the intermediate calculations are large in size.
+All operators are supported, including sorting, aggregating, joining, and windowing. 
+Future work can further test the boundaries of what is possible with DuckDB's out-of-core support, including window functions and even larger data sizes. 
+
+## Conclusion
+
+In summary, not only is DuckDB's feature set growing substantially with each release, DuckDB is getting faster very fast! 
+Overall, performance has improved by **14 times in only 3 years!**
+
+Yet query performance is only part of the story!
+The variety of workloads that DuckDB can handle is wide and growing wider thanks to a full-featured SQL dialect, including high performance window functions.
+Additionally, critical workloads like data import, CSV parsing, and data export have improved dramatically over time.
+The complete developer experience is critical for DuckDB!
+
+Finally, DuckDB now supports larger-than-memory calculations across all operators: sorting, aggregating, joining, and windowing.
+The size of problem that you can handle on your current compute resources just got 10x bigger, or more! 
+
+If you have made it this far, welcome to the flock! 🦆
+[Join us on Discord](https://discord.duckdb.org/), we value your feedback!
+
```

Comment:
Can the data be exported as a duckdb database file, and saved somewhere?

So people can run their own queries on it?
We have a blobs.duckdb.org to host random datasets, this can easily be one of the

This implies some sort of commitment to update it, so maybe do save the scripts used in some repo somewhere.

Reply:
Ah, we can do that! I can provide the output as both a SQLite file and a DuckDB file (I used SQLite so that I could check on the progress of the benchmark while it was running in a separate process). Right now I have them in a few separate files, especially for the scale one since it kept running my laptop out of hard disk space... ;-). I will consolidate them.

How do I upload to blobs.duckdb.org?

I should save the scripts - good point. The results analysis are in a pair of Jupyter notebooks and they ended up getting pretty messy with all the combinations of visualizations that I tried. The script for running the benchmark is better, but I'd love a code review on that as well (but maybe should add some comments first...).